doc/sphinx/proof-engine/tactics.rst

```diff
@@ -650,180 +668,381 @@ Applying theorems
+   (The ``forall``\s may also be interleaved with the premises, but common usage is
+   to equivalently gather them at the beginning of the :n:`@one_term`.)
+   Backward reasoning with a :n:`@one_term` whose type is, for example, `A -> B`
+   replaces an as-yet unproven goal `B` with `A`.  Forward reasoning with the same
+   :n:`@one_term` changes a hypothesis with type `A` to `B`.  (Hypotheses are
+   considered proven propositions within the context that contains them.)
+
+   Unification creates a map from the variables in the type of :n:`@one_term`
+   to matching subterms of the goal or hypothesis.
+   The matching subterms are then substituted into the type of :n:`@one_term`
+   when generating the updated goal or hypothesis.  Unmatched premises become
+   new subgoals with similar substitutions.  If no match is found, the
+   tactic fails.
+
+   The goal and hypothesis cases are described separately for clarity.
+
+.. _unused1:
+
+   .. the dummy ref name is needed to get correct formatting of the next line and "Without..."
+
+   :n:`@one_term` (inside :n:`@one_term_with_bindings`)
+     If :n:`@one_term` is an :n:`@ident`, it is the name of
+     a theorem, lemma or hypothesis whose type is given in the
+     theorem statement or shown in the context.  Otherwise it is a proof term whose
+     type can be displayed with :cmd:`Check` :n:`@one_term`.
+
+   Without :n:`@in_hyp_as` (the goal case)
+     If the goal matches all of the type of :n:`@one_term` (both premises and
+     the conclusion), the tactic proves the goal.
+     Otherwise, the tactic matches the goal against the conclusion of :n:`@one_term`
+     and, if possible, one or more premises (from right to left).
+     If the match succeeds, the tactic replaces the current goal with a subgoal for
+     each unmatched premise of the type of :n:`@one_term`.  This
+     :ref:`example <apply_backward>` matches only the conclusion, while
+     this :ref:`one <apply_backward_w_premises>` also matches a premise.
+
+     If the conclusion of the type of :token:`one_term` does not match the goal
+     *and* the conclusion is an inductive type with a single constructor,
+     then each premise in the constructor is recursively matched to the goal in
+     left-to-right order and the first match is used.  In this case, the tactic
+     will not match premises that would result in applying a lemma of the form
+     ``forall A, … -> A``.  See example :ref:`here <apply_with_iff>`.
+
+.. _apply_with_second_order_unification:
+
+     The goal case uses first-order unification with dependent types unless the
+     conclusion of the type of :token:`term` is of the form
+     :n:`P (t__1 … t__n)` with ``P`` to be instantiated. In the latter case,
+     the behavior depends on the form of the target. If the target is of the form
+     :n:`(fun x__1 … x__n => Q) u__1 … u__n` and the :n:`t__i` and :n:`u__i` unify,
+     then :g:`P` is instantiated into :n:`(fun x__1 … x__n => Q)`. Otherwise,
```

Comment:
After reading this again and comparing it to the behavior I observed, I'm convinced that the meaning of this description is not that the target has this form syntactically, but only semantically (a function applied to arguments). So to make this clear, we should not use the `fun` syntactic construct in this part of the description (only later when we talk of using `pattern`).
```suggestion
     the behavior depends on the form of the target. If the target is of the form
     :n:`Q u__1 … u__n` and the :n:`t__i` and :n:`u__i` unify,
     then :n:`P` is instantiated into :n:`Q`. Otherwise,
```